### PR TITLE
mihomo-party: init at 1.4.5

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes
- Init mihomo-party at 1.4.5, Another Mihomo GUI, https://github.com/mihomo-party-org/mihomo-party

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/mi/mihomo-party/package.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+          libGL
+        ]
+      }"
+  '';
+
+  meta = {
+    description = "Another Mihomo GUI";
+    homepage = "https://github.com/mihomo-party-org/mihomo-party";
+    mainProgram = "mihomo-party";
+    platforms = [ "x86_64-linux" ];
```
Comment (h7x4) [inline pkgs/by-name/mi/mihomo-party/package.nix:73]:
```suggestion
    platforms = with platforms; linux ++ darwin ++ windows;
    broken = !(stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isx86_64);
```

File: pkgs/by-name/mi/mihomo-party/package.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    license = lib.licenses.gpl3;
```
Comment (h7x4) [inline pkgs/by-name/mi/mihomo-party/package.nix:74]:
```suggestion
    license = lib.licenses.gpl3Plus;
```

File: pkgs/by-name/mi/mihomo-party/package.nix [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    libayatana-appindicator
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir $out
+    cp -r opt/mihomo-party usr/share $out
+    substituteInPlace $out/share/applications/mihomo-party.desktop \
+      --replace "/opt/mihomo-party/mihomo-party" "mihomo-party"
```
Comment (h7x4) [inline pkgs/by-name/mi/mihomo-party/package.nix:54]:
```suggestion
      --replace-fail "/opt/mihomo-party/mihomo-party" "mihomo-party"
```

File: pkgs/by-name/mi/mihomo-party/package.nix [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  udev,
+  libayatana-appindicator,
+  libGL,
+}:
+let
+  pname = "mihomo-party";
+  version = "1.4.0";
+  src = fetchurl {
+    url = "https://github.com/mihomo-party-org/mihomo-party/releases/download/v${version}/mihomo-party-linux-${version}-amd64.deb";
+    sha256 = "/iC63QXEQ8DnrHTkgYO6Audxib4GqScDIk76kDypN6E=";
```
Comment (h7x4) [inline pkgs/by-name/mi/mihomo-party/package.nix:22]: Could you recalculate this as an SRI hash?

```suggestion
    hash = "sha256-????";
```

**Review by h7x4 — APPROVED** [review]

**Comment by h7x4** [comment]
I'm having some troubles testing the binary. Maybe it just doesn't support my GPU? First error just seems to be that I have misconfigured my file picker.

```
[973551:1001/085449.536140:ERROR:object_proxy.cc(576)] Failed to call method: org.freedesktop.DBus.Properties.Get: object_path= /org/freedesktop/portal/desktop: org.freedesktop.DBus.Error.InvalidArgs: No such interface “org.freedesktop.portal.FileChooser”
[973551:1001/085449.536170:ERROR:select_file_dialog_linux_portal.cc(287)] Failed to read portal version property
ac: Unknown GPU, using 0 for raster_config
Cannot find target for triple amdgcn-- Unable to find target for this triple (no targets are registered)
(node:973551) UnhandledPromiseRejectionWarning: Error: No such file or directory (os error 2)
    at disableSysProxy (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:1311:14)
    at triggerSysProxy (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:1255:13)
    at App.<anonymous> (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:4221:3)
    at App.emit (node:events:519:28)
    at /nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:4237:20
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
(Use `mihomo-party --trace-warnings ...` to show where the warning was created)
(node:973551) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
[973551:1001/085507.076378:ERROR:gpu_process_host.cc(1007)] GPU process exited unexpectedly: exit_code=139

┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
     Sub-Store -- v2.14.385
┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅

[sub-store] INFO: Start migrating...
[sub-store] INFO: Migration complete!
[sub-store] INFO: [BACKEND] :::38324
[973551:1001/085507.078404:ERROR:network_service_instance_impl.cc(608)] Network service crashed, restarting service.
ac: Unknown GPU, using 0 for raster_config
Cannot find target for triple amdgcn-- Unable to find target for this triple (no targets are registered)
```

**Comment by aucub** [comment]
> I'm having some troubles testing the binary. Maybe it just doesn't support my GPU? First error just seems to be that I have misconfigured my file picker.
> 
> ```
> [973551:1001/085449.536140:ERROR:object_proxy.cc(576)] Failed to call method: org.freedesktop.DBus.Properties.Get: object_path= /org/freedesktop/portal/desktop: org.freedesktop.DBus.Error.InvalidArgs: No such interface “org.freedesktop.portal.FileChooser”
> [973551:1001/085449.536170:ERROR:select_file_dialog_linux_portal.cc(287)] Failed to read portal version property
> ac: Unknown GPU, using 0 for raster_config
> Cannot find target for triple amdgcn-- Unable to find target for this triple (no targets are registered)
> (node:973551) UnhandledPromiseRejectionWarning: Error: No such file or directory (os error 2)
>     at disableSysProxy (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:1311:14)
>     at triggerSysProxy (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:1255:13)
>     at App.<anonymous> (/nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:4221:3)
>     at App.emit (node:events:519:28)
>     at /nix/store/m8gfw2004sz4b4az2rmzry91pw9kdqij-mihomo-party-1.4.0/mihomo-party/resources/app.asar/out/main/index.js:4237:20
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> (Use `mihomo-party --trace-warnings ...` to show where the warning was created)
> (node:973551) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
> [973551:1001/085507.076378:ERROR:gpu_process_host.cc(1007)] GPU process exited unexpectedly: exit_code=139
> 
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
>      Sub-Store -- v2.14.385
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
> 
> [sub-store] INFO: Start migrating...
> [sub-store] INFO: Migration complete!
> [sub-store] INFO: [BACKEND] :::38324
> [973551:1001/085507.078404:ERROR:network_service_instance_impl.cc(608)] Network service crashed, restarting service.
> ac: Unknown GPU, using 0 for raster_config
> Cannot find target for triple amdgcn-- Unable to find target for this triple (no targets are registered)
> ```

These issues do not occur on my device (Vega 6/GNOME 46), so it may be worth reporting upstream. I also found some related issues:
- [Spotube Issue #1904](https://github.com/KRTirtho/spotube/issues/1904)
- [VSCode Issue #226391](https://github.com/microsoft/vscode/issues/226391)

**Comment by h7x4** [comment]
I tried running the application on a different machine, with X11 and a different graphics card. It failed with another error, `应用初始化失败: Error: no such file or directory (os error 2)`.
```
[nix-shell:~/.cache/nixpkgs-review/pr-345189]$ ./results/mihomo-party/bin/mihomo-party
(node:105663) UnhandledPromiseRejectionWarning: Error: No such file or directory (os error 2)
    at disableSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1312:14)
    at triggerSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1256:13)
    at App.<anonymous> (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4226:3)
    at App.emit (node:events:519:28)
    at /nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4242:20
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
(Use `mihomo-party --trace-warnings ...` to show where the warning was created)
(node:105663) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)

┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
     Sub-Store -- v2.14.385
┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅

[sub-store] INFO: [BACKEND] :::38324
```

strace seems to suggest it's looking for some flatpak related stuff through a relative path:

```
[nix-shell:~/.cache/nixpkgs-review/pr-345189/results/mihomo-party/bin]$ strace -e trace=file ./mihomo-party
...
access("/.flatpak-info", F_OK)          = -1 ENOENT (No such file or directory)
```

**Comment by aucub** [comment]
> I tried running the application on a different machine, with X11 and a different graphics card. It failed with another error, `应用初始化失败: Error: no such file or directory (os error 2)`.我尝试在另一台机器上运行该应用程序，使用 X11 和不同的显卡。它失败了，并出现另一个错误 `应用初始化失败: Error: no such file or directory (os error 2)`。
> 
> ```
> [nix-shell:~/.cache/nixpkgs-review/pr-345189]$ ./results/mihomo-party/bin/mihomo-party
> (node:105663) UnhandledPromiseRejectionWarning: Error: No such file or directory (os error 2)
>     at disableSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1312:14)
>     at triggerSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1256:13)
>     at App.<anonymous> (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4226:3)
>     at App.emit (node:events:519:28)
>     at /nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4242:20
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> (Use `mihomo-party --trace-warnings ...` to show where the warning was created)
> (node:105663) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
> 
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
>      Sub-Store -- v2.14.385
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
> 
> [sub-store] INFO: [BACKEND] :::38324
> ```
> 
> strace seems to suggest it's looking for some flatpak related stuff through a relative path:strace 似乎表明它正在通过相对路径查找一些与 flatpak 相关的东西：
> 
> ```
> [nix-shell:~/.cache/nixpkgs-review/pr-345189/results/mihomo-party/bin]$ strace -e trace=file ./mihomo-party
> ...
> access("/.flatpak-info", F_OK)          = -1 ENOENT (No such file or directory)
> ```

If you can provide your device's system information, that would be helpful. The application may have limited support for desktop environments other than KDE or Gnome.

### First Issue:
In the `disableSysProxy` function, the library `@mihomo-party/sysproxy` (https://www.npmjs.com/package/@mihomo-party/sysproxy) is used. This library relies on `sysproxy-rs` (https://github.com/mihomo-party-org/sysproxy-rs/blob/main/src/linux.rs) for its implementation. Based on an analysis of `sysproxy-rs`, when running on non-KDE desktop environments, the application uses the Gnome-specific settings. Obviously, this could lead to errors on desktop environments other than KDE or Gnome. It seems that error handling has been implemented in the parent function `triggerSysProxy` in the code.

### Second Issue:
The absence of the `.flatpak-info` file is expected behavior and occurs even when the application is functioning correctly. The application searches for these files to check if it is running in an AppImage or Flatpak package. Therefore, this cannot be classified as an error.

Based on the current information, it is unclear whether these two issues are related to the application's failure to start. The first issue appears to be unrelated to the second. I suggest reporting the startup issue upstream for further investigation.

**Comment by aucub** [comment]
> I tried running the application on a different machine, with X11 and a different graphics card. It failed with another error, `应用初始化失败: Error: no such file or directory (os error 2)`.
> 
> ```
> [nix-shell:~/.cache/nixpkgs-review/pr-345189]$ ./results/mihomo-party/bin/mihomo-party
> (node:105663) UnhandledPromiseRejectionWarning: Error: No such file or directory (os error 2)
>     at disableSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1312:14)
>     at triggerSysProxy (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:1256:13)
>     at App.<anonymous> (/nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4226:3)
>     at App.emit (node:events:519:28)
>     at /nix/store/c4hmw1sfl533dn117asncqaqk09kfpyw-mihomo-party-1.4.1/mihomo-party/resources/app.asar/out/main/index.js:4242:20
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> (Use `mihomo-party --trace-warnings ...` to show where the warning was created)
> (node:105663) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
> 
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
>      Sub-Store -- v2.14.385
> ┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅┅
> 
> [sub-store] INFO: [BACKEND] :::38324
> ```
> 
> strace seems to suggest it's looking for some flatpak related stuff through a relative path:
> 
> ```
> [nix-shell:~/.cache/nixpkgs-review/pr-345189/results/mihomo-party/bin]$ strace -e trace=file ./mihomo-party
> ...
> access("/.flatpak-info", F_OK)          = -1 ENOENT (No such file or directory)
> ```

If you can provide your device's system information, that would be helpful. The application may have limited support for desktop environments other than KDE or Gnome.

### First Issue:
In the `disableSysProxy` function, the library `@mihomo-party/sysproxy` (https://www.npmjs.com/package/@mihomo-party/sysproxy) is used. This library relies on `sysproxy-rs` (https://github.com/mihomo-party-org/sysproxy-rs/blob/main/src/linux.rs) for its implementation. Based on an analysis of `sysproxy-rs`, when running on non-KDE desktop environments, the application uses the Gnome-specific settings. Obviously, this could lead to errors on desktop environments other than KDE or Gnome. It seems that error handling has been implemented in the parent function `triggerSysProxy` in the code.

### Second Issue:
The absence of the `.flatpak-info` file is expected behavior and occurs even when the application is functioning correctly. The application searches for these files to check if it is running in an AppImage or Flatpak package. Therefore, this cannot be classified as an error.

Based on the current information, it is unclear whether these two issues are related to the application's failure to start. The first issue appears to be unrelated to the second. I suggest reporting the startup issue upstream for further investigation.

**Comment by aucub** [comment]
@h7x4
I installed Hyprland on my device to test the application, and encountered the same issue: `应用初始化失败: Error: no such file or directory (os error 2)`. After commenting out `triggerAutoProxy` and `triggerManualProxy` in the code, the application started successfully.

You can replace the link with [this release](https://github.com/aucub/mihomo-party/releases/download/v1.4.1/mihomo-party-linux-1.4.1-amd64.deb) to retest, or try launching the application using **nix-alien**.

**Comment by h7x4** [comment]
Sorry for the wait. I've retested it, works well :+1:

**Review by Zaalaanko — COMMENTED** [review]
Parake mesirbe pueden tomar mis datos pribados

**Pull request merged**